cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,1849 @@ func TestCreateVPCSubnet(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		vpcSecurityGroupStatus := make(map[string]infrav1beta2.VPCSecurityGroupStatus)
+		vpcSecurityGroupStatus[securityGroupName] = infrav1beta2.VPCSecurityGroupStatus{ID: &securityGroupID, RuleIDs: []*string{ptr.To("ruleID")}, ControllerCreated: ptr.To(true)}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPCSecurityGroups: vpcSecurityGroupStatus,
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, infrav1beta2.VPCSecurityGroup{
+						Name: ptr.To("securityGroupName"),
+					}),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroup(gomock.Any()).Return(&vpcv1.SecurityGroup{ID: ptr.To("securityGroupID"), Name: ptr.To("securityGroupName")}, nil, nil)
+		mockVPC.EXPECT().GetSecurityGroupRule(gomock.Any()).Return(nil, nil, errors.New("failed to get security group rule"))
+		err := clusterScope.ReconcileVPCSecurityGroups()
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When SecurityGroup is created successfully but returns error while creating SecurityGroupRules", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		rules := infrav1beta2.VPCSecurityGroupRule{
+			Direction: infrav1beta2.VPCSecurityGroupRuleDirectionInbound,
+			Source: &infrav1beta2.VPCSecurityGroupRulePrototype{
+				Remotes: append([]infrav1beta2.VPCSecurityGroupRuleRemote{}, infrav1beta2.VPCSecurityGroupRuleRemote{
+					Address:    ptr.To("192.168.0.1/24"),
+					RemoteType: infrav1beta2.VPCSecurityGroupRuleRemoteTypeAddress,
+				}),
+				Protocol: infrav1beta2.VPCSecurityGroupRuleProtocolTCP,
+				PortRange: &infrav1beta2.VPCSecurityGroupPortRange{MaximumPort: 65535,
+					MinimumPort: 1,
+				},
+			},
+		}
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("securityGroupID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, infrav1beta2.VPCSecurityGroup{
+						Name:  ptr.To("securityGroupName"),
+						Rules: append([]*infrav1beta2.VPCSecurityGroupRule{}, &rules),
+					}),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroupByName(gomock.Any()).Return(nil, nil)
+		mockVPC.EXPECT().CreateSecurityGroup(gomock.Any()).Return(&vpcv1.SecurityGroup{ID: ptr.To("securityGroupID")}, nil, nil)
+		mockVPC.EXPECT().CreateSecurityGroupRule(gomock.Any()).Return(nil, nil, errors.New("failed to create security group rule"))
+		err := clusterScope.ReconcileVPCSecurityGroups()
+		g.Expect(err).ToNot(BeNil())
+	})
+}
+
+func TestValidateVPCSecurityGroup(t *testing.T) {
```

Comment:
Can you test this case too?
`VPC security group by name exists but is not attached to VPC`

Reply:
have added ut for this 
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/2011/files#diff-04d9f443c4026a9cd5e26cf630140aa890cf62140d697655d0293a8d97970f57R3438